### ISSUE TITLE
Tracking Issue for `const_is_char_boundary`

### ISSUE THREAD

**Issue by zachs18** [rust-lang/rust#131516]
Feature gate: `#![feature(const_is_char_boundary)]`

This is a tracking issue for using `str::is_char_boundary` in `const`, which allows checking that `index`-th byte is the first byte in a UTF-8 code point sequence or the end of the string during const-eval.

### Public API

```rust
// core::str
impl str {
    pub const fn is_char_boundary(&self, index: usize) -> bool;
}
```

### Steps / History

- [x] Implementation: #131520
- [ ] Final comment period (FCP)[^1]
- [ ] Stabilization PR

### Unresolved Questions

- None yet.

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

**Comment by okaneco** [comment]
I was looking at this as well, it opens the door for making `str::split_at*` and the `round_char_boundary` functions const (the latter needs the iterator code to be rewritten).

I don't believe the check for `index == 0` is necessary so one possible implementation is as follows. 
```rust
pub const fn is_char_boundary(&self, index: usize) -> bool {
    if index < self.as_bytes().len() {
        self.as_bytes()[index].is_utf8_char_boundary()
    } else {
        // The start and end of a string are considered boundaries
        index == self.len()
    }
}
```

**Comment by zachs18** [comment]
Yeah, this was motivated by making `str::split_at*` const (cc #131518, which #131520 also implements). I looked at `{floor,ceil}_char_boundary` but did not want to mess with them at the moment since the rest were relatively simple changes.
